File: IrrgartenP3/IrrgartenJava/src/irrgarten/Labyrinth.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 *
 * @author moham
 */// Labyrinth.java
import java.util.ArrayList;

public class Labyrinth {
    private static final char BLOCK_CHAR_OBSTACLE = 'X';
    private static final char BLOCK_CHAR_EMPTY = '-';
    private static final char BLOCK_CHAR_MONSTER = 'M';
    private static final char BLOCK_CHAR_EXIT = 'E';
    private static final char BLOCK_CHAR_COMBAT = 'C';

    private static final int ROW = 0;
    private static final int COL = 1;

    private int nRows, nCols;
    private int exitRow, exitCol; // Exit coordinates
    private Monster[][] monsters;
    private Player[][] players;
    private char[][] cells;

    public Labyrinth(int nRows, int nCols) {
        this.nRows = nRows;
        this.nCols = nCols;
        this.monsters = new Monster[nRows][nCols];
        this.players = new Player[nRows][nCols];
        this.cells = new char[nRows][nCols];

        // Initialize all cells to empty
        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nCols; j++) {
                cells[i][j] = BLOCK_CHAR_EMPTY;
            }
        }

        // Set initial exit position to an invalid location
        this.exitRow = -1;
        this.exitCol = -1;
    }

    public void setExit(int row, int col){
        this.exitRow = row;
        this.exitCol = col;
        cells[exitRow][exitCol] = BLOCK_CHAR_EXIT;

    }
    
    public boolean haveAWinner() {
        // Check if any player is at the exit position
        return players[exitRow][exitCol] != null;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nCols; j++) {
                sb.append(cells[i][j]).append(' ');
            }
            sb.append('\n');
        }
        return sb.toString();
    }

    public void addMonster(int row, int col, Monster monster) {
        if (posOK(row, col) && cells[row][col] == BLOCK_CHAR_EMPTY) {
            monsters[row][col] = monster;
            monster.setPos(row, col);
            cells[row][col] = BLOCK_CHAR_MONSTER;
        }
    }

    public boolean posOK(int row, int col) {
        return row >= 0 && row < nRows && col >= 0 && col < nCols;
    }

    public boolean emptyPos(int row, int col) {
        return cells[row][col] == BLOCK_CHAR_EMPTY;
    }

    public boolean monsterPos(int row, int col) {
        return cells[row][col] == BLOCK_CHAR_MONSTER;
    }

    public boolean exitPos(int row, int col) {
        return row == exitRow && col == exitCol;
    }

    public boolean combatPos(int row, int col) {
        return cells[row][col] == BLOCK_CHAR_COMBAT;
    }

    public boolean canStepOn(int row, int col) {
        return posOK(row, col) && (emptyPos(row, col) || monsterPos(row, col) || exitPos(row, col));
    }

    public void updateOldPos(int row, int col) {
        if (posOK(row, col)) {
            if (combatPos(row, col)) {
                cells[row][col] = BLOCK_CHAR_MONSTER;
            } else {
                cells[row][col] = BLOCK_CHAR_EMPTY;
            }
        }
    }

    public int[] dir2Pos(int row, int col, Directions direction) {
        int[] newPos = {row, col};
        switch (direction) {
            case UP: newPos[ROW]--; break;
            case DOWN: newPos[ROW]++; break;
            case LEFT: newPos[COL]--; break;
            case RIGHT: newPos[COL]++; break;
        }
        return newPos;
    }

    public int[] randomEmptyPos() {
        int[] position;
        do {
            int row = Dice.randomPos(nRows);
            int col = Dice.randomPos(nCols);
            position = new int[]{row, col};
        } while (!emptyPos(position[ROW], position[COL]));
        return position;
    }

    // Dentro de la clase Labyrinth
    public void spreadPlayers(ArrayList<Player> players) {
        for (Player p : players) {
            // Obtener una posición vacía aleatoria
            int[] pos = randomEmptyPos();
            // Colocar al jugador en la nueva posición
            putPlayer2D(-1, -1, pos[ROW], pos[COL], p);
        }
    }


    public Monster putPlayer(Directions direction, Player player) {
        // 1.1: Obtener la posición actual del jugador
        int oldRow = player.getRow();
        int oldCol = player.getCol();

        // 1.3: Obtener la nueva posición en función de la dirección
        int[] newPos = dir2Pos(oldRow, oldCol, direction);

        // 1.4: Mover al jugador a la nueva posición y devolver el monstruo en la nueva posición si lo hay
        return putPlayer2D(oldRow, oldCol, newPos[ROW], newPos[COL], player);
    }


    public void addBlock(Orientation orientation, int startRow, int startCol, int length) {
        int incRow = 0, incCol = 0;

        // Definir incrementos según la orientación
        if (orientation == Orientation.VERTICAL) {
            incRow = 1;
        } else {
            incCol = 1;
        }

        int row = startRow, col = startCol;

        // Colocar los bloques en las posiciones correspondientes
        while (posOK(row, col) && emptyPos(row, col) && length > 0) {
            cells[row][col] = BLOCK_CHAR_OBSTACLE;
            length--;
            row += incRow;
            col += incCol;
        }
    }


    public ArrayList<Directions> validMoves(int row, int col) {
        ArrayList<Directions> output = new ArrayList<>();

        // Agregar direcciones válidas en función de las posiciones libres
        if (canStepOn(row + 1, col)) output.add(Directions.DOWN);
        if (canStepOn(row - 1, col)) output.add(Directions.UP);
        if (canStepOn(row, col + 1)) output.add(Directions.RIGHT);
        if (canStepOn(row, col - 1)) output.add(Directions.LEFT);

        return output;
    }


    public Monster putPlayer2D(int oldRow, int oldCol, int row, int col, Player player) {
        Monster output = null;

        // Verificar si la nueva posición es válida para moverse
        if (canStepOn(row, col)) {
            // Si la posición antigua es válida, actualizarla a vacío
            if (posOK(oldRow, oldCol)) {
                Player p = this.players[oldRow][oldCol];
                if(p == player){
                    updateOldPos(oldRow, oldCol);
                    this.players[oldRow][oldCol] = null;
                }
            }

            boolean monsterPos = this.monsterPos(row, col);
            
            // Verificar si la nueva posición tiene un monstruo
            if (monsterPos) {
                cells[row][col] = BLOCK_CHAR_COMBAT;
                output = monsters[row][col];
            } else {
                cells[row][col] = player.getNumber();
            }

            // Actualizar la nueva posición del jugador en el tablero
            this.players[row][col] = player;
            player.setPos(row, col);
        }

        return output;
    }
    
    public void removeMonster(Monster monster, Player winnerPlayer) {
        int row = monster.getRow();
        int col = monster.getCol();

        // Verificar si la posición es válida y contiene el monstruo
        if (posOK(row, col) && monsters[row][col] == monster) {
            monsters[row][col] = null;       // Eliminar el monstruo de la matriz
            cells[row][col] = winnerPlayer.getNumber();  // Actualizar la celda para reflejar que está vacía
        }
    }

}

File: IrrgartenP4/irrgartenRuby/irrgarten/labyrinth.rb
module Irrgarten
  class Labyrinth
    BLOCK_CHAR_OBSTACLE = 'X'
    BLOCK_CHAR_EMPTY = '-'
    BLOCK_CHAR_MONSTER = 'M'
    BLOCK_CHAR_EXIT = 'E'
    BLOCK_CHAR_COMBAT = 'C'

    ROW = 0
    COL = 1

    attr_reader :n_rows, :n_cols, :exit_row, :exit_col

    def initialize(n_rows, n_cols)
      @n_rows = n_rows
      @n_cols = n_cols
      @monsters = Array.new(n_rows) { Array.new(n_cols) }
      @players = Array.new(n_rows) { Array.new(n_cols) }
      @cells = Array.new(n_rows) { Array.new(n_cols, BLOCK_CHAR_EMPTY) }
      @exit_row = -1
      @exit_col = -1
    end

    def set_exit(row, col)
      @exit_row = row
      @exit_col = col
      @cells[row][col] = BLOCK_CHAR_EXIT
    end

    def have_a_winner?
      !@players[@exit_row][@exit_col].nil?
    end

    def to_s
      @cells.map { |row| row.join(' ') }.join("\n")
    end

    def add_monster(row, col, monster)
      if pos_ok?(row, col) && @cells[row][col] == BLOCK_CHAR_EMPTY
        @monsters[row][col] = monster
        monster.set_pos(row, col)
        @cells[row][col] = BLOCK_CHAR_MONSTER
      end
    end

    def pos_ok?(row, col)
      row.between?(0, @n_rows - 1) && col.between?(0, @n_cols - 1)
    end

    def empty_pos?(row, col)
      @cells[row][col] == BLOCK_CHAR_EMPTY
    end

    def monster_pos?(row, col)
      @cells[row][col] == BLOCK_CHAR_MONSTER
    end

    def exit_pos?(row, col)
      row == @exit_row && col == @exit_col
    end

    def combat_pos?(row, col)
      @cells[row][col] == BLOCK_CHAR_COMBAT
    end

    def can_step_on?(row, col)
      pos_ok?(row, col) && (empty_pos?(row, col) || monster_pos?(row, col) || exit_pos?(row, col))
    end

    def update_old_pos(row, col)
      if pos_ok?(row, col)
        @cells[row][col] = combat_pos?(row, col) ? BLOCK_CHAR_MONSTER : BLOCK_CHAR_EMPTY
      end
    end

    def dir_to_pos(row, col, direction)
      case direction
      when :up then [row - 1, col]
      when :down then [row + 1, col]
      when :left then [row, col - 1]
      when :right then [row, col + 1]
      else [row, col]
      end
    end

    def random_empty_pos
      loop do
        row = Dice.random_pos(@n_rows)
        col = Dice.random_pos(@n_cols)
        return [row, col] if empty_pos?(row, col)
      end
    end

    def spread_players(players)
      players.each do |player|
        pos = random_empty_pos
        put_player2d(-1, -1, pos[ROW], pos[COL], player)
      end
    end

    def put_player(direction, player)
      old_row = player.row
      old_col = player.col

      new_pos = dir_to_pos(old_row, old_col, direction)

      put_player2d(old_row, old_col, new_pos[ROW], new_pos[COL], player)
    end

    def add_block(orientation, start_row, start_col, length)
      inc_row, inc_col = 0, 0

      # Set increments based on orientation
      if orientation == :vertical
        inc_row = 1
      else
        inc_col = 1
      end

      row, col = start_row, start_col

      # Place blocks in the specified direction
      while pos_ok?(row, col) && empty_pos?(row, col) && length > 0
        @cells[row][col] = BLOCK_CHAR_OBSTACLE
        length -= 1
        row += inc_row
        col += inc_col
      end
    end

    def valid_moves(row, col)
      output = []

      output << :down if can_step_on?(row + 1, col)
      output << :up if can_step_on?(row - 1, col)
      output << :right if can_step_on?(row, col + 1)
      output << :left if can_step_on?(row, col - 1)

      output
    end

    def put_player2d(old_row, old_col, new_row, new_col, player)
      return nil unless can_step_on?(new_row, new_col)

      # Update old position to empty if it's valid
      if pos_ok?(old_row, old_col) && @players[old_row][old_col] == player
        update_old_pos(old_row, old_col)
        @players[old_row][old_col] = nil
      end

      # Check if the new position contains a monster
      if monster_pos?(new_row, new_col)
        @cells[new_row][new_col] = BLOCK_CHAR_COMBAT
        monster = @monsters[new_row][new_col]
      else
        @cells[new_row][new_col] = player.number
        monster = nil
      end

      # Update the player's new position on the board
      @players[new_row][new_col] = player
      player.set_pos(new_row, new_col)

      monster
    end
    
    def replace_player(old_player, new_player)
      @players.each_with_index do |row, r_idx|
        row.each_with_index do |cell, c_idx|
          if cell == old_player
            @players[r_idx][c_idx] = new_player
            new_player.set_pos(r_idx, c_idx)
            break
          end
        end
      end
    end
    
  end
end